### ISSUE TITLE
Azure Raporlama

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#430]
<html>
<body>
<html><head></head><body><h1>Raporlama mimarisi — log-centric, Azure üzerinde (şema + bileşenler + kısa açıklama)</h1>
<p>Aşağıdaki mimari, <strong>mevcut log dosyalarını</strong> (senin örnek: <code inline="">file:///mnt/data/live_trading_20251125_150553_419453.log</code>) doğrudan okuyup raporlayan, üretim için ölçeklenebilir ve maliyet-dengeli bir çözüm sunar. Öncelik: <em>sıfır/az kod değişikliği</em> → parser tabanlı ETL → Azure ingestion → Power BI + e-posta raporu.</p>
<hr>
<h2>#1 — Yüksek seviyeli diyagram (metinsel)</h2>
<pre><code>[Azure VM: Docker (bearish-alpha-bot)]
    └─ Writes log files to /var/log/bot/*.log
         (example file: file:///mnt/data/live_trading_20251125_150553_419453.log)
             |
             v
[Log Parser Service - Docker container on same VM]
    - tails new logs OR runs periodically
    - extracts structured JSON events (ndjson)
    - writes to: /data/parsed/events/&lt;run_id&gt;.ndjson
             |
             v
[Fluent Bit sidecar on VM] 
    - tail /data/parsed/*.ndjson
    - output -&gt; Azure Log Analytics (or Event Hubs)
             |
             v
[Azure Log Analytics Workspace]  &lt;-----------------------------
    - short-term logs, alerts                                   |
    - continuous export -&gt; Azure Data Explorer (ADX)            |
             |                                                  |
             v                                                  |
[Azure Data Explorer (ADX)]  --&gt; Power BI (DirectQuery)        |
    - optimized for interactive analytics, retention, KQL      |
             |                                                  |
             v                                                  |
[Power BI Service]  &lt;--- (dashboards, scheduled refresh)       |
    - detailed dashboards (equity curve, trade table, ML hist) |
             ^                                                  |
             |                                                  |
[Azure Function] &lt;--- Logic App (trigger on Container Stop /   |
    - when Logic App triggers with run_id/file info:            |
      • query Log Analytics / ADX for run events (KQL)          |
      • compute KPIs, render HTML, convert to PDF              |
      • store report in Blob Storage (reports/{run_id}.pdf)    |
      • send e-mail (SendGrid/Office365) with SAS link          |
                                                               |
[Azure Blob Storage] -- stores raw logs (daily), reports, ndjson
                                                               |
[Azure Monitor / Alerts] &lt;------------------------------------- 
    - Alert rules (container stop, daily loss, high exception rate)
    - Action Groups -&gt; Logic App / Email / SMS
                                                               |
[Key Vault &amp; Managed Identity]  - credentials &amp; secure access
</code></pre>
<hr>
<h2>#2 — Hangi Azure servisleri neden kullanılmalı (liste)</h2>
<ul>
<li>
<p><strong>Azure VM (mevcut)</strong> — bot ve parser container çalışır.</p>
</li>
<li>
<p><strong>Fluent Bit (VM sidecar)</strong> — parsed NDJSON dosyalarını topluca Azure Log Analytics veya Event Hubs’a göndermek için hafif ve güvenilir.</p>
</li>
<li>
<p><strong>Azure Log Analytics Workspace</strong> — kısa dönem central logging; kolay KQL sorguları ve Alerts için kullanılır.</p>
</li>
<li>
<p><strong>Container Insights (Azure Monitor)</strong> — container lifecycle (start/stop/restart) ve temel container metrikleri (CPU/RAM/restarts). Container stop event’i için Logic App tetikleyicisi olarak kullan.</p>
</li>
<li>
<p><strong>Azure Data Explorer (ADX)</strong> — Power BI ile düşük gecikmeli, büyük ölçekli analitik için (continuous export ile Log Analytics → ADX). ADX, cumulative PnL, expectancy gibi hızlı sorgular için ideal.</p>
</li>
<li>
<p><strong>Azure Blob Storage</strong> — raw logs, parsed ndjson dosyaları ve üretilmiş raporların (PDF) arşivlenmesi (Parquet tercih edilebilir).</p>
</li>
<li>
<p><strong>Azure Functions</strong> — KQL sorgularını çalıştırıp KPI hesaplama, HTML şablon oluşturma, PDF üretimi (headless Chromium) ve Blob’a upload; idempotency kontrolü.</p>
</li>
<li>
<p><strong>Logic Apps</strong> — tetikleyiciler ve orkestrasyon (Container Insights event veya Log Analytics query alert) — Function çağırır.</p>
</li>
<li>
<p><strong>Power BI Service</strong> — ADX’ten DirectQuery ile dashboard’lar (equity curve, trades table, ML_confidence hist).</p>
</li>
<li>
<p><strong>Azure Monitor Alerts &amp; Action Groups</strong> — uyarılar (daily loss, exception spike, container disconnect) → Logic App/Function/Teams/email.</p>
</li>
<li>
<p><strong>Azure Key Vault + Managed Identity</strong> — tüm secret’ların korunması ve servisler arası kimlik doğrulama.</p>
</li>
<li>
<p><strong>Optional: Event Hubs</strong> — çok yüksek event hacmi veya ileri pipeline gerekiyorsa (ör. Databricks), NDJSON’ı önce Event Hubs’a yönlendir.</p>
</li>
</ul>
<hr>
<h2>#3 — Veri akışı (adım adım, run örneği)</h2>
<ol>
<li>
<p>Bot tamamlandığında veya çalışırken log dosyasına kayıt yapar; dosya örneği:<br>
<code inline="">file:///mnt/data/live_trading_20251125_150553_419453.log</code></p>
</li>
<li>
<p><strong>Log Parser Service</strong> (Docker): bu dosyayı okur, satırları regex/semantic rules ile parse eder ve <code inline="">event</code> nesneleri üretir (ndjson). Her event içinde <code inline="">run_id</code> alanı olmalı (varsa log’dan al, yoksa filename timestamp kullan).</p>
</li>
<li>
<p>Parser çıktısını bir klasöre yazar: <code inline="">/data/parsed/&lt;run_id&gt;.ndjson</code>.</p>
</li>
<li>
<p><strong>Fluent Bit</strong> tails bu ndjson dosyasını ve gönderir → <strong>Log Analytics</strong> (veya Event Hubs).</p>
</li>
<li>
<p><strong>Log Analytics</strong>: kısa süreli sorgulama ve alert için temel kayıt. Buna ek olarak <strong>continuous export</strong> ile veriler <strong>ADX</strong>’e akar.</p>
</li>
<li>
<p><strong>Logic App</strong> tespit eder: Container Insights’ın stop event’i veya Log Analytics’te alert (ör. <code inline="">shutdown</code> eventi bulundu) → Logic App çağırır <strong>Azure Function</strong> ile <code inline="">run_id</code>.</p>
</li>
<li>
<p><strong>Function</strong> ADX/Log Analytics’ten run_id ile eventleri çeker (KQL), KPI hesaplar, HTML rapor üretir; raporu PDF’ye çevirir, Blob’a yükler ve e-posta gönderir (SendGrid/Office365). Idempotency: Function, <code inline="">reports/{run_id}.pdf</code> var ise tekrar üretmez, sadece mevcut linki döndürür.</p>
</li>
<li>
<p><strong>Power BI</strong> ADX’e bağlanır (DirectQuery) — detaylı dashboard gerçek zamanlı/near-real time görüntüler.</p>
</li>
</ol>
<hr>
<h2>#4 — Örnek tablo şeması (ADX / Log Analytics mapping)</h2>
<p><strong>ADX tablo: <code inline="">bearish_events</code></strong> (nginx JSON fields mapped)</p>

Kolon adı | Tip | Açıklama
-- | -- | --
run_id | string | run id (from log filename or field)
timestamp_utc | datetime | event time
event_type | string | run_start, model_prediction, signal_generated, trade_entry, trade_exit, shutdown, exception, health_ping, log
logger | string | logger name
level | string | INFO/WARNING/ERROR
message | string | raw message
symbol | string | BTC/USDT (if trade event)
entry_price | real | (if trade_entry)
exit_price | real | (if trade_exit)
pnl_usd | real | (if trade_exit)
holding_time_s | long |  
strategy | string | strategy name
ml_confidence | real | if available
rl_confidence | real | if available
signal_score | real |  
extra | dynamic | JSON blob for vendor-specific fields

<hr>
<h2>#5 — Kısa Power BI &amp; KQL örnekleri</h2>
<ul>
<li>
<p><strong>Toplam trade / gün</strong></p>
</li>
</ul>
<pre><code class="language-kql">bearish_events
| where event_type == "trade_exit" and timestamp_utc &gt;= ago(1d)
| summarize total_trades = count(), total_pnl = sum(pnl_usd)
</code></pre>
<ul>
<li>
<p><strong>Expectancy (30 gün)</strong></p>
</li>
</ul>
<pre><code class="language-kql">let t = bearish_events | where event_type == "trade_exit" and timestamp_utc &gt;= ago(30d);
let wins = toscalar(t | where pnl_usd &gt; 0 | summarize cnt = count()); 
let losses = toscalar(t | where pnl_usd &lt;= 0 | summarize cnt = count());
let avg_win = toscalar(t | where pnl_usd &gt; 0 | summarize avg = avg(pnl_usd));
let avg_loss = toscalar(t | where pnl_usd &lt;= 0 | summarize avg = avg(pnl_usd));
print expectancy = avg_win*(wins/(wins+losses)) - abs(avg_loss)*(losses/(wins+losses))
</code></pre>
<ul>
<li>
<p><strong>Cumulative PnL (grafik)</strong></p>
</li>
</ul>
<pre><code class="language-kql">bearish_events
| where event_type == "trade_exit"
| order by timestamp_utc asc
| project timestamp_utc, pnl_usd
| extend cum_pnl = row_cumsum(pnl_usd)
</code></pre>
<hr>
<h2>#6 — Güvenlik, idempotency ve operasyonel notlar</h2>
<ul>
<li>
<p><strong>Idempotency:</strong> Function rapor üretmeden önce <code inline="">reports/{run_id}.pdf</code> kontrolü yapmalı; varsa yeniden üretme.</p>
</li>
<li>
<p><strong>Secrets:</strong> APPINSIGHTS_CONN, Log Analytics workspace key, Storage conn string, SendGrid key → <strong>Key Vault</strong>. Function ve VM için <strong>Managed Identity</strong> kullan.</p>
</li>
<li>
<p><strong>Retention policy:</strong> Log Analytics kısa (30 gün), ADX orta (90–365 gün), raw logs Blob: 1+ yıl (parquet).</p>
</li>
<li>
<p><strong>Cost control:</strong> ADX ingestion ve Log Analytics sorgu maliyetlerini izlemek için Azure Budget kur.</p>
</li>
<li>
<p><strong>Fail-safe:</strong> Eğer ADX anlık kullanılmazsa Function direkt Log Analytics’e sorgu atabilmelidir (geçici fallback).</p>
</li>
</ul>
<hr>
<h2>#7 — Minimal bileşen listesi (ilk deploy için, P0 → hızlı kurulum)</h2>
<ol>
<li>
<p>Existing Azure VM (Docker) — bot already running</p>
</li>
<li>
<p>Add <strong>Log Parser</strong> container on same VM (Docker Compose kullan)</p>
</li>
<li>
<p>Install <strong>Fluent Bit</strong> on VM (as container or service) — tail parsed ndjson</p>
</li>
<li>
<p>Create <strong>Log Analytics Workspace</strong> (if not exists)</p>
</li>
<li>
<p>Enable <strong>Container Insights</strong> for VM</p>
</li>
<li>
<p>Configure <strong>Fluent Bit output</strong> to Log Analytics (or Event Hubs)</p>
</li>
<li>
<p>Create <strong>Logic App</strong> (Container Stop trigger) → call <strong>Azure Function</strong></p>
</li>
<li>
<p>Create <strong>Azure Function</strong> skeleton (query Log Analytics, generate report, upload to Blob, email)</p>
</li>
<li>
<p>Create <strong>Blob Storage</strong> container <code inline="">reports/</code></p>
</li>
<li>
<p>Create <strong>Power BI</strong> dashboard (connect to ADX when available; initially you can test with exported CSV)</p>
</li>
</ol>
<hr>
<h2>#8 — Nasıl başlıyoruz (önerilen ilk 3 adım, hızlı)</h2>
<ol>
<li>
<p><strong>Deploy Log Parser container</strong> (kodu ben hazırlarım) — parser, <code inline="">file:///mnt/data/live_trading_20251125_150553_419453.log</code> dosyasını okuyup <code inline="">/data/parsed/&lt;run_id&gt;.ndjson</code> üretir. (Test: run parser locally.)</p>
</li>
<li>
<p><strong>Install Fluent Bit</strong> on VM and configure output to Log Analytics (provide workspace id &amp; key). Test: parsed ndjson events appear in Log Analytics.</p>
</li>
<li>
<p><strong>Create Logic App + Function skeleton</strong> — Logic App configured to trigger on container stop; function receives <code inline="">run_id</code> and returns "OK". Test: manually call Logic App with sample <code inline="">run_id</code>.</p>
</li>
</ol>
<hr>
<h3>Ek not</h3>
<p>Log dosyası örneğinizin yolu (test / kaynak):<br>
<code inline="">file:///mnt/data/live_trading_20251125_150553_419453.log</code> — bu dosyayı parser’a giriş olarak kullanacağız.</p>
<hr>

</body>
</html>
